service/stopserving/stopserving.go

```diff
@@ -0,0 +1,141 @@
+	FetchBucketsLimit = int64(500)
+	// FetchBucketsInterval define the interval to fetch buckets for stop serving
+	FetchBucketsInterval = 5 * time.Minute
+	// SubmitTransactionInterval defines the interval between two stop serving transaction submissions
+	SubmitTransactionInterval = 500 * time.Millisecond
+)
+
+// StopServing module is responsible for stop serving buckets on testnet.
+type StopServing struct {
+	config   *StopServingConfig
+	cache    *lru.Cache
+	running  atomic.Value
+	stopCh   chan struct{}
+	signer   *signerclient.SignerClient
+	metadata *metadataclient.MetadataClient
+}
+
+// NewStopServingService returns an instance of stop serving
+func NewStopServingService(cfg *StopServingConfig) (*StopServing, error) {
+	var (
+		stopServing *StopServing
+		err         error
+	)
+
+	stopServing = &StopServing{
+		config: cfg,
+		stopCh: make(chan struct{}),
+	}
+	if stopServing.cache, err = lru.New(model.LruCacheLimit); err != nil {
+		log.Errorw("failed to create lru cache", "error", err)
+		return nil, err
+	}
+	if stopServing.signer, err = signerclient.NewSignerClient(cfg.SignerGrpcAddress); err != nil {
+		log.Errorw("failed to create signer client", "error", err)
+		return nil, err
+	}
+	if stopServing.metadata, err = metadataclient.NewMetadataClient(cfg.MetadataGrpcAddress); err != nil {
+		log.Errorw("failed to create metadata client", "error", err)
+		return nil, err
+	}
+	log.Debugw("stop serving service created successfully")
+	return stopServing, nil
+}
+
+// Name return the stop serving service name
+func (s *StopServing) Name() string {
+	return model.StopServingService
+}
+
+// Start function start background goroutine to execute stop serving
+func (s *StopServing) Start(ctx context.Context) error {
+	if s.running.Swap(true) == true {
+		return errors.New("stop serving has already started")
+	}
+
+	if s.config.DiscontinueConfig.BucketKeepAliveDays > 0 {
+		// start background task
+		go s.eventLoop()
+	}
+
+	return nil
+}
+
+// eventLoop a background goroutine to periodically conduct stop serving
+func (s *StopServing) eventLoop() {
+	s.discontinueBuckets()
+	ticker := time.NewTicker(FetchBucketsInterval)
+	for {
+		select {
+		case <-ticker.C:
+			s.discontinueBuckets()
+		case <-s.stopCh:
+			return
+		}
+	}
+}
+
+// discontinueBuckets fetch buckets from metadata service and submit transactions to chain
+func (s *StopServing) discontinueBuckets() {
+	createAt := time.Now().Unix() - int64(86400*s.config.DiscontinueConfig.BucketKeepAliveDays)
+	buckets, err := s.metadata.ListExpiredBucketsBySp(context.Background(), createAt, s.config.SpOperatorAddress, FetchBucketsLimit)
+	if err != nil {
+		log.Errorw("failed to query expired buckets", "error", err)
+		return
+	}
+
+	for _, bucket := range buckets {
+		bucketName := bucket.BucketInfo
+		if s.cache.Contains(bucketName) { // this bucket has been discontinued, however the metadata indexer did not handle it yet
```

Comment:
Is it better to use BucketID ?

Reply:
And For long-term thinking, we may use the same cache for discontinueBucket and discontinueObject. So, the key will has a prefix, E.g b{BucketID}, o{ObjectID}

Reply:
refactor to use the format you mentioned